### PR TITLE
builder: Initial git support

### LINKED ISSUE #25: Implement git as an upstream source
Here are my ideas on how this will work:

# Directory Layout
Inside the git cache directory (`/var/cache/boulder/upstreams/git`), different sources will be split into different folders similar to how Go downloads modules (e.g. `github.com/serpent-os/moss` -> `/var/cache/boulder/upstreams/git/github.com/serpent-os/moss`, `invent.kde.org/graphics/krita` -> `/var/cache/boulder/upstreams/git/invent.kde.org/graphics/krita`).

# Running git
The `libgit2` bindings for D seem to be [inactive for a long time](https://github.com/s-ludwig/dlibgit), so I would consider just running the plain `git` command. The process of fetching would look like this:
1. Parse the corresponding directory to store the source (`/var/cache/boulder/upstreams/git/github.com/serpent-os/moss`) and check if it exists already. If it doesn't, go to step 2 directly. If it does,
    1. `git restore . && git clean -xf .` to remove any unwanted extra files if the build files were able to get into this directory. Even if they can't, I think it's still worth running this just to be safe.
    2. `git fetch` to pull in any latest commits and refs. This also serves as a check that this repository upstream is still valid and reachable after the last build.
    3. If any of the above step fails, then report the user about it, delete the directory, and continue as if nothing happened.
2. `git clone --recursive-submodules ...`
3. `git checkout $REQUESTED_REF`
4. The rest will be the same as the plain source.

I'd be glad to implement this if accepted.

### PR DESCRIPTION
**Pull request by GZGavinZhao**
Fixes #25. Depends on serpent-os/moss-fetcher#1 and serpent-os/moss-core#8.

~~Will remove some tracing code used for debugging once finished testing.~~ Done.

I will update the commit message later. Code should be ready for review. Thanks for the wait!

### CODE REVIEW

**Comment by ermo** [comment]
Please be sure to run `serpent-style/setup.sh` to enable git commit hooks.

In addition, please look for statements that include format strings and ensure that they are checked at compile time:

GOOD: `infof!"some message with some format string %s"(someformatstring);`

BAD: `infof("some message with some format string %", someformatstring);`

The overall goal is to ensure that ALL format strings are checked at compile time, in order to remove them as a source of potential runtime errors.

File: source/boulder/upstreamcache.d [hunk 1]
```diff
@@ -106,25 +215,73 @@ public final class UpstreamCache
...
+        else if (def.type == UpstreamType.Git)
+        {
+            import std.path;
+            import std.algorithm.searching : startsWith;
+
+            string path = join([gitDirectory, normalizedUriPath(def.uri)], "/");
+            trace("This git source's finalPath will be " ~ path);
+
+            /* A very simple check to prevent path escaping */
+            enforce(!path.asRelativePath(gitDirectory).startsWith(".."),
```
Comment (ikeycode) [inline source/boulder/upstreamcache.d:275]: for some reason i really like this :)

Comment (GZGavinZhao) [reply source/boulder/upstreamcache.d:284]: :D

**Comment by ikeycode** [comment]
I'm glad the caveat is accounted for and documented re the submodules. Thanks!! Awesome work

**Pull request merged**